src/index.js

```diff
@@ -1,212 +1,169 @@
+      if (
+        // Block code
+        // Check from https://github.com/leafac/rehype-shiki
+        node.tagName === 'pre' &&
+        Array.isArray(node.children) &&
+        node.children.length === 1 &&
+        node.children[0].tagName === 'code' &&
+        typeof node.children[0].properties === 'object' &&
+        Array.isArray(node.children[0].properties.className) &&
+        typeof node.children[0].properties.className[0] === 'string' &&
+        node.children[0].properties.className[0].startsWith('language-')
+      ) {
+        const codeNode = node.children[0].children[0];
+        const lang = node.children[0].properties.className[0].split('-')[1];
+        const meta =
+          node.children[0].data?.meta ?? node.children[0].properties.metastring;
+        const lineNumbers = meta
+          ? rangeParser(meta.match(/{(.*)}/)?.[1] ?? '')
+          : [];
+        const word = meta?.match(/\/(.*)\//)?.[1];
+        const wordNumbers = meta
+          ? rangeParser(meta.match(/\/.*\/([^\s]*)/)?.[1] ?? '')
+          : [];
+
+        const trees = {};
+        for (const [mode, highlighter] of highlighterCache.entries()) {
+          trees[mode] = hastParser.parse(
+            highlighter.codeToHtml(codeNode.value, lang)
+          );
         }
 
-        if (word && node.innerHTML.includes(word)) {
-          visitedWordsCount++;
-
-          if (
-            wordNumbers.length === 0 ||
-            wordNumbers.includes(visitedWordsCount)
-          ) {
-            const splitByBreakChar = new RegExp(`\\b${word}\\b`);
-            const adjacentText = node.innerHTML.split(splitByBreakChar);
-
-            node.innerHTML = adjacentText
-              .map(
-                (txt, i) =>
-                  `${txt}${
-                    i !== adjacentText.length - 1
-                      ? `<span data-mdx-pretty-code-word>${word}</span>`
-                      : ''
-                  }`
-              )
-              .join('');
-
-            node
-              .querySelectorAll('[data-mdx-pretty-code-word]')
-              .forEach((wordNode) => {
-                wordNode.removeAttribute('data-mdx-pretty-code-word');
-                onVisitHighlightedWord(wordNode);
+        Object.entries(trees).forEach(([mode, tree]) => {
+          let lineCounter = 0;
+          let wordCounter = 0;
+
+          visit(tree, 'element', (node) => {
+            if (node.properties.className?.[0] === 'line') {
+              onVisitLine(node);
+
+              if (
+                lineNumbers.length !== 0 &&
+                lineNumbers.includes(++lineCounter)
+              ) {
+                onVisitHighlightedLine(node);
+              }
+
+              // TODO: handle words that span across syntax boundaries/nodes.
+              // https://github.com/atomiks/mdx-pretty-code/issues/3
+              node.children.forEach((child) => {
+                if (child.children[0].value.includes(word)) {
```

Comment:
This should be easier to implement now, probably by looping over the nodes in a given line and add them together, seeing if they match